server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/StreamCompletionListener.java

```diff
@@ -0,0 +1,88 @@
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.stream.read.listener;
+
+import org.opensearch.common.io.Channels;
+import org.opensearch.common.io.InputStreamContainer;
+import org.opensearch.core.action.ActionListener;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class StreamCompletionListener implements ActionListener<Void> {
+    private final int partNumber;
+    private final InputStreamContainer blobPartStreamContainer;
+    private final Path fileLocation;
+    private final AtomicBoolean anyPartStreamFailed;
+    private final FileCompletionListener fileCompletionListener;
+
+    // 8 MB buffer for transfer
+    private static final int BUFFER_SIZE = 8 * 1024 * 2024;
+
+    public StreamCompletionListener(
+        int partNumber,
+        InputStreamContainer blobPartStreamContainer,
+        Path fileLocation,
+        AtomicBoolean anyPartStreamFailed,
+        FileCompletionListener fileCompletionListener
+    ) {
+        this.partNumber = partNumber;
+        this.blobPartStreamContainer = blobPartStreamContainer;
+        this.fileLocation = fileLocation;
+        this.anyPartStreamFailed = anyPartStreamFailed;
+        this.fileCompletionListener = fileCompletionListener;
+    }
+
+    @Override
+    public void onResponse(Void unused) {
+        // Ensures no writes to the file if any stream fails.
+        if (!anyPartStreamFailed.get()) {
+            try (FileChannel outputFileChannel = FileChannel.open(fileLocation, StandardOpenOption.WRITE, StandardOpenOption.CREATE)) {
+                try (InputStream inputStream = blobPartStreamContainer.getInputStream()) {
+                    outputFileChannel.position(blobPartStreamContainer.getOffset());
+
+                    final byte[] buffer = new byte[BUFFER_SIZE];
+                    ByteBuffer byteBuffer = ByteBuffer.wrap(buffer);
+                    int bytesRead;
+
+                    while ((bytesRead = inputStream.read(buffer)) != -1) {
+                        byteBuffer.limit(bytesRead);
+                        Channels.writeToChannel(byteBuffer, outputFileChannel);
+                        byteBuffer.clear();
+                    }
+                }
+            } catch (IOException e) {
+                onFailure(e);
+                return;
+            }
+            fileCompletionListener.onResponse(partNumber);
+        }
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        try {
+            if (Files.exists(fileLocation)) {
+                Files.delete(fileLocation);
+            }
+        } catch (IOException ex) {
+            // Die silently
+        }
+        if (!anyPartStreamFailed.get()) {
+            anyPartStreamFailed.compareAndSet(false, true);
```

Comment:
I don't think this is thread safe. What I think you want is this:

```
if (anyPartStreamFailed.getAndSet(true) == false) {
    fileCompletionListener.onFailure(e);
}
```

On the first caller to flip the value from false to true will invoke the listener, all other calls end up being no-ops.